pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,184 @@
+    _collectionsToUnnest = new ArrayList<>(unnestCollections);
+    Collections.sort(_collectionsToUnnest);
+  }
+
+  public static boolean isComplexTypeHandlingEnabled(TableConfig tableConfig) {
+    if (tableConfig.getIngestionConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode() == null) {
+      return false;
+    }
+    return tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode()
+        != ComplexTypeHandlingConfig.Mode.NONE;
+  }
+
+  @Nullable
+  @Override
+  public GenericRow transform(GenericRow record) {
+    flattenMap(record, new HashSet<>(record.getFieldToValueMap().keySet()));
+    for (String collection : _collectionsToUnnest) {
+      unnestCollection(record, collection);
+    }
+    return record;
+  }
+
+  private GenericRow unnestCollection(GenericRow record, String column) {
+    if (record.getValue(GenericRow.MULTIPLE_RECORDS_KEY) == null) {
+      List<GenericRow> list = new ArrayList<>();
+      unnestCollection(record, column, list);
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    } else {
+      Collection<GenericRow> records = (Collection) record.getValue(GenericRow.MULTIPLE_RECORDS_KEY);
+      List<GenericRow> list = new ArrayList<>();
+      for (GenericRow innerRecord : records) {
+        unnestCollection(innerRecord, column, list);
+      }
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    }
+    return record;
+  }
+
+  private void unnestCollection(GenericRow record, String column, List<GenericRow> list) {
+    Object value = record.removeValue(column);
+    if (value == null) {
+      // use the record itself
+      list.add(record);
+      return;
+    } else if (value instanceof Collection) {
+      if (((Collection) value).isEmpty()) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Collection) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    } else if (value.getClass().isArray()) {
+      if (((Object[]) value).length == 0) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Object[]) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    }
+  }
+
+  private GenericRow flattenCollectionItem(GenericRow record, Object obj, String column) {
+    GenericRow copy = record.copy();
+    if (obj instanceof Map) {
+      Map<String, Object> map = (Map<String, Object>) obj;
+      for (Map.Entry<String, Object> entry : new HashSet<>(map.entrySet())) {
+        String flattenName = concat(column, entry.getKey());
+        copy.putValue(flattenName, entry.getValue());
+      }
+    } else {
+      copy.putValue(column, obj);
+    }
+    return copy;
```

Comment:
Since we are not recursing here, looks like list inside a list is not handled. What if obj is a list ?

Example 

```
"addresses": [
        {
          "apt": 1012,
          "zip": 94404,
          "phones": [
            {
              "phonenumber": 123456,
              "type": "mobile"
            },
            {
              "phonenumber": 565676,
              "type": "work"
            },
            {
              "phonenumber": 212123244,
              "type": "emergency"
            }
          ]
        },
        {
          "apt": 1022,
          "zip": 94402,
          "phones": [
            {
              "phonenumber": 212121,
              "type": "mobile"
            },
            {
              "phonenumber": 32322,
              "type": "work"
            },
            {
              "phonenumber": 232424,
              "type": "landline"
            }
          ]
        }
      ]
```

addresses = list (struct (apt, city, zip, struct (phones (list (struct ( phone-number, type ))))))

The above code won't flatten the phones list inside the addresses list.
 
The above code will be called for each item in the root list addresses. Each item is a struct (in-memory hashmap). We go over the map assuming each entry is non-list.

Reply:
it does, in the code of
```
for (String collection : _collectionsToUnnest) {
      unnestCollection(record, collection);
    }
```
where _collectionsToUnnest is sorted (topologically). There is a test case for this

Reply:
If we take the json snippet above as an example, then what would the final output of transform(GenericRow record) function (output rows, columns, values) will look like?

Also, will the final output contain only the leaf nodes of the tree or also intermediate nodes?

Reply:
Can you take a look at https://github.com/apache/incubator-pinot/pull/6845/files#diff-fc3e3bcb113fc2349533e59dca933233903769ab670153f2fd255e2a56877cc1R78 and see if the tests are sufficient?